core/src/main/java/org/sonarsource/sonarlint/core/repository/rules/RulesRepository.java

```diff
@@ -26,45 +26,62 @@
 import java.util.stream.Collectors;
 import javax.annotation.CheckForNull;
 import org.sonarsource.sonarlint.core.rule.extractor.SonarLintRuleDefinition;
+import org.sonarsource.sonarlint.core.rules.RulesExtractionHelper;
 
 public class RulesRepository {
+
+  private final RulesExtractionHelper extractionHelper;
   private Map<String, SonarLintRuleDefinition> embeddedRulesByKey;
   private final Map<String, Map<String, SonarLintRuleDefinition>> rulesByKeyByConnectionId = new HashMap<>();
   private final Map<String, Map<String, String>> ruleKeyReplacementsByConnectionId = new HashMap<>();
 
-  public void setEmbeddedRules(Collection<SonarLintRuleDefinition> embeddedRules) {
-    this.embeddedRulesByKey = byKey(embeddedRules);
+  public RulesRepository(RulesExtractionHelper extractionHelper) {
+    this.extractionHelper = extractionHelper;
   }
 
-  @CheckForNull
   public Collection<SonarLintRuleDefinition> getEmbeddedRules() {
-    return embeddedRulesByKey == null ? null : embeddedRulesByKey.values();
+    lazyInit();
+    return embeddedRulesByKey.values();
   }
 
   public Optional<SonarLintRuleDefinition> getEmbeddedRule(String ruleKey) {
+    lazyInit();
     return Optional.ofNullable(embeddedRulesByKey.get(ruleKey));
   }
 
-  @CheckForNull
+  private synchronized void lazyInit() {
+    if (embeddedRulesByKey == null) {
+      this.embeddedRulesByKey = byKey(extractionHelper.extractEmbeddedRules());
+    }
+  }
+
   public Collection<SonarLintRuleDefinition> getRules(String connectionId) {
+    lazyInit(connectionId);
+    return rulesByKeyByConnectionId.getOrDefault(connectionId, Map.of()).values();
+  }
+
+  public Optional<SonarLintRuleDefinition> getRule(String connectionId, String ruleKey) {
+    lazyInit(connectionId);
+    var connectionRules = rulesByKeyByConnectionId.get(connectionId);
+    return Optional.ofNullable(connectionRules.get(ruleKey))
```

Comment:
Potential NPE here on `connectionRules` ? And I see it's just code that was moved